torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -30,22 +169,48 @@ class ChunkDataSet : public BatchDataset<Self, Batch, size_t> {
 
   using SelfType = Self;
   using BatchType = Batch;
-  using BatchRequestType = size_t;
   using ChunkSamplerType = ChunkSampler;
   using ExampleSamplerType = ExampleSampler;
 
+  ChunkDataSet(size_t preloader_count) : preloader_count_(preloader_count) {
+    if (preloader_count_ == 0) {
+      throw std::runtime_error(
+          "preloader_count is 0. At least one preloader needs to be specified.");
+    }
+    chunk_sampler_ =
+        torch::make_unique<samplers::ThreadSafeSampler<ChunkSamplerType>>(
+            ChunkSamplerType(0));
+  }
+
+  /// Customize copy constructor. We don't expect user to copy ChunkDataSet
+  /// which inner state. This is error-prone if user changes some internal state
+  /// while DataLoader is using it. Instead, we create a new ChunkDataSet from a
+  /// clean slate, and DataLoader uses shared pointer to reference this
+  /// ChunkDataSet.
+  ChunkDataSet(const ChunkDataSet& data_set)
+  {
+    preloader_count_ = data_set.preloader_count_;
+    chunks_to_load_ = 3;
+    chunk_sampler_ =
+        torch::make_unique<samplers::ThreadSafeSampler<ChunkSamplerType>>(
+            chunks_to_load_);
+    lazy_initialize();
+  }
+
   virtual ~ChunkDataSet() {
-    // stop prefetching chunks.
+    for (auto& worker_thread : preload_threads_) {
+      worker_thread.join();
```

Comment:
Join the threads. Blocking the current thread until the workers finish execution. 
We can optimize this a bit by setting a boolean to break the infinite loop in worker, which makes them exit earlier, i.e. less blocking. The draw back of using the Boolean is that, we need to add a lock every time we read/write this value, which could be a regular additional overhead. Given that currently the blocking is from the process of tearing down object, I think we can live with current design.